clever/src/led.cpp

```diff
@@ -0,0 +1,300 @@
+	} else if (!set_leds.response.success) {
+		ROS_WARN_THROTTLE(5, "Calling set_leds failed: %s", set_leds.response.message.c_str());
+	}
+}
+
+void rainbow(uint8_t n, uint8_t& r, uint8_t& g, uint8_t& b)
+{
+	if (n < 255 / 3) {
+		r = n * 3;
+		g = 255 - n * 3;
+		b = 0;
+	} else if (n < 255 / 3 * 2) {
+		n -= 255 / 3;
+		r = 255 - n * 3;
+		g = 0;
+		b = n * 3;
+	} else {
+		n -= 255 / 3 * 2;
+		r = 0;
+		g = n * 3;
+		b = 255 - n * 3;
+	}
+}
+
+void fill(uint8_t r, uint8_t g, uint8_t b)
+{
+	set_leds.request.leds.resize(led_count);
+	for (int i = 0; i < led_count; i++) {
+		set_leds.request.leds[i].index = i;
+		set_leds.request.leds[i].r = r;
+		set_leds.request.leds[i].g = g;
+		set_leds.request.leds[i].b = b;
+	}
+	call_set_leds();
+}
+
+void proceed(const ros::TimerEvent& event)
+{
+	counter++;
+	uint8_t r, g, b;
+	set_leds.request.leds.clear();
+	set_leds.request.leds.resize(led_count);
+
+	if (current_effect.effect == "blink" || current_effect.effect == "blink_fast") {
+		blink_state = !blink_state;
+		// toggle all leds
+		if (blink_state) {
+			fill(current_effect.r, current_effect.g, current_effect.b);
+		} else {
+			fill(0, 0, 0);
+		}
+
+	} else if (current_effect.effect == "fade") {
+		// fade all leds from starting state
+		double passed = std::min((event.current_real - start_time).toSec() / fade_period, 1.0);
+		double one_minus_passed = 1 - passed;
+		for (int i = 0; i < led_count; i++) {
+			set_leds.request.leds[i].index = i;
+			set_leds.request.leds[i].r = one_minus_passed * start_state.leds[i].r + passed * current_effect.r;
+			set_leds.request.leds[i].g = one_minus_passed * start_state.leds[i].g + passed * current_effect.g;
+			set_leds.request.leds[i].b = one_minus_passed * start_state.leds[i].b + passed * current_effect.b;
+		}
+		call_set_leds();
+		if (passed >= 1.0) {
+			// fade finished
+			timer.stop();
+		}
+
+	} else if (current_effect.effect == "rainbow_fill") {
+		rainbow(counter % 255, r, g, b);
+		for (int i = 0; i < led_count; i++) {
+			set_leds.request.leds[i].index = i;
+			set_leds.request.leds[i].r = r;
+			set_leds.request.leds[i].g = g;
+			set_leds.request.leds[i].b = b;
+		}
+		call_set_leds();
+
+	} else if (current_effect.effect == "rainbow") {
+		for (int i = 0; i < led_count; i++) {
+			int pos = (int)round(counter + (255.0 * i / led_count)) % 255;
+			rainbow(pos % 255, r, g, b);
+			set_leds.request.leds[i].index = i;
+			set_leds.request.leds[i].r = r;
+			set_leds.request.leds[i].g = g;
+			set_leds.request.leds[i].b = b;
+		}
+		call_set_leds();
+	}
+}
+
+bool setEffect(clever::SetLEDEffect::Request& req, clever::SetLEDEffect::Response& res)
+{
+	res.success = true;
+	counter = 0;
+
+	if (req.effect != "flash" && current_effect.effect == req.effect &&
+	    current_effect.r == req.r && current_effect.g == req.g && current_effect.b == req.b) {
+		res.message = "Effect already set, skip";
+		return true;
+	}
+
+	if (req.effect == "" || req.effect == "fill") {
```

Comment:
Wouldn't this be better as a `switch`? You'd need to convert your string to something like an enum, but that would be more readable in the long run in my opinion.

Also, you should use named constants for your strings instead of raw strings - this would mean less chance of bugs due to typos.

Reply:
Personally, I don't any benefits giver by using `switch`. Conversely, I see disadvantages, about possibility of forgotten `break` and falling though.

Reply:
There's a `-Wimplicit-fallthrough` flag for that since gcc 7. Still, at the very least please convert your string literals to constants (although I still don't think multiple `if` statements with string comparisons are considered good C++ style).

Reply:
Okay, let's just hope this won't be too hard to maintain.